lib/vm/src/instance.rs

```diff
@@ -688,6 +688,189 @@ impl Instance {
+
+        // We need to do some pointer arithmetic now. The unit is `u8`.
+        let ptr = self.instance_ptr.cast::<u8>().as_ptr();
+        let base_ptr = ptr.add(mem::size_of::<Instance>());
+
+        for i in 0..num_memories {
+            let mem_offset = self
+                .offsets
+                .vmctx_vmmemory_definition(LocalMemoryIndex::new(i));
+            let mem_offset = usize::try_from(mem_offset).unwrap();
+
+            let new_ptr = NonNull::new_unchecked(base_ptr.add(mem_offset));
+
+            out.push(new_ptr.cast());
+        }
+
+        out
+    }
+
+    /// Get the locations of where the [`VMTableDefinition`]s should be stored.
+    ///
+    /// This function lets us create [`Table`] objects on the host with backing
+    /// memory in the VM.
+    ///
+    /// # Safety
+    /// - `instance_ptr` must point to enough memory that all of the
+    ///   offsets in `offsets` point to valid locations in memory,
+    ///   i.e. `instance_ptr` must have been allocated by
+    ///   `Self::new`.
+    unsafe fn table_definition_locations(&self) -> Vec<NonNull<VMTableDefinition>> {
+        let num_tables = self.offsets.num_local_tables;
+        let num_tables = usize::try_from(num_tables).unwrap();
+        let mut out = Vec::with_capacity(num_tables);
+
+        // We need to do some pointer arithmetic now. The unit is `u8`.
+        let ptr = self.instance_ptr.cast::<u8>().as_ptr();
+        let base_ptr = ptr.add(std::mem::size_of::<Instance>());
+
+        for i in 0..num_tables {
+            let table_offset = self
+                .offsets
+                .vmctx_vmtable_definition(LocalTableIndex::new(i));
+            let table_offset = usize::try_from(table_offset).unwrap();
+
+            let new_ptr = NonNull::new_unchecked(base_ptr.add(table_offset));
+
+            out.push(new_ptr.cast());
+        }
+        out
+    }
+
+    /// Finish preparing by writing the [`Instance`] into memory.
+    pub(crate) fn write_instance(mut self, instance: Instance) -> InstanceAllocator {
+        unsafe {
+            // `instance` is moved at `instance_ptr`. This pointer has
+            // been allocated by `Self::allocate_instance` (so by
+            // `InstanceAllocator::allocate_instance`.
+            ptr::write(self.instance_ptr.as_ptr(), instance);
+            // Now `instance_ptr` is correctly initialized!
+        }
+        let instance = self.instance_ptr;
+        let instance_layout = self.instance_layout;
+        // prevent the old state's drop logic from being called as we
+        // transition into the new state.
+        self.consumed = true;
+
+        // This is correct because of the invariants of `Self` and
+        // because we write `Instance` to the pointer in this function.
+        unsafe { InstanceAllocator::new(instance, instance_layout) }
+    }
+
+    /// Get the [`VMOffsets`] for the allocated buffer.
+    pub(crate) fn offsets(&self) -> VMOffsets {
+        self.offsets.clone()
```

Comment:
Why cloning? Do we want a “read-only” API? In any case, I think returning `&self.offsets` should be OK. Thoughts?